src/fragment-generation-utils.js

```diff
@@ -1141,6 +1152,47 @@ const getLastNodeForBlockSearch = (range) => {
   return node;
 };
 
+/**
+ * Finds the first visible text node within a given range.
+ * @param {Range} range - range in which to find the first visible text node
+ * @returns {Node} - first visible text node within |range| or null if there are
+ * no visible text nodes within |range|
+ */
+const getFirstTextNode = (range) => {
+  // Check if first node in the range is a visible text node.
+  const firstNode = getFirstNodeForBlockSearch(range);
+  if (isText(firstNode) && fragments.internal.isNodeVisible(firstNode)) {
+    return firstNode;
+  }
+
+  // First node is not visible text, use a tree walker to find the first visible
+  // text node.
+  const walker = fragments.internal.makeTextNodeWalker(range);
+  walker.currentNode = firstNode;
+
+  return walker.nextNode();
+};
+
+/**
+ * Finds the last visible text node within a given range.
+ * @param {Range} range - range in which to find the last visible text node
+ * @returns {Node} - last visible text node within |range| or null if there are
+ * no visible text nodes within |range|
+ */
+const getLastTextNode = (range) => {
+  // Check if last node in the range is a visible text node.
+  const lastNode = getLastNodeForBlockSearch(range);
+  if (isText(lastNode) && fragments.internal.isNodeVisible(lastNode)) {
+    return lastNode;
+  }
+
+  // Last node is not visible text, traverse the range backwards to find the
+  // last visible text node.
+  const walker = fragments.internal.makeTextNodeWalker(range);
```

Comment:
Should we set `walker.currentNode = lastNode` here?

Reply:
That makes total sense, both there and on the start side. I missed that even when the walker will emit only nodes in the range, it will still have to traverse all the nodes from the root to the star/end container of the range before it could pass the filter function. Starting on the first/last node would save us that extra traversal.